File: core/src/main/java/at/shockbytes/corey/common/core/util/view/SliderPreference.java
package at.shockbytes.corey.common.core.util.view;

import android.app.AlertDialog;
import android.content.Context;
import android.os.Bundle;
import android.preference.DialogPreference;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Button;
import android.widget.SeekBar;
import android.widget.SeekBar.OnSeekBarChangeListener;
import android.widget.TextView;

import at.shockbytes.corey.common.core.R;


/**
 * This is an own implementation of the SliderPreference, which allows the user
 * to choose between an interval from 0 - x seconds. This interval is the
 * location update interval in the TrackingService class.
 * 
 * @author Martin Macheiner
 *
 */
public class SliderPreference extends DialogPreference implements
		OnSeekBarChangeListener, OnClickListener {

	private static final String androidns = "http://schemas.android.com/apk/res/android";

	private SeekBar mSeekBar;
	private TextView mValueText;

	private String mUnit;

	private int mDefault;
	private int mMax;
	private int mValue;

	/**
	 * @param context
	 *            Application context
	 * @param attrs
	 *            AttributeSet from xml definition
	 */
	public SliderPreference(Context context, AttributeSet attrs) {
		super(context, attrs);
		initialize(attrs);
	}

	/**
	 * @param context
	 *            Application context
	 * @param attrs
	 *            AttributeSet from xml definition
	 * @param defStyleAttr
	 *            Integer with Styling
	 */
	public SliderPreference(Context context, AttributeSet attrs,
			int defStyleAttr) {
		super(context, attrs, defStyleAttr);
		initialize(attrs);
	}

	/**
	 * Initializes all values from AttributeSet
	 *
	 * @param _attrs
	 *            AttributeSet from constructor
	 */
	private void initialize(AttributeSet _attrs) {

		//Get string value for suffix
		int unitId = _attrs.getAttributeResourceValue(androidns, "txtExercise", 0);
		mUnit = (unitId == 0) ? _attrs.getAttributeValue(androidns, "txtExercise")
				: getContext().getString(unitId);

		mDefault = _attrs.getAttributeIntValue(androidns, "defaultValue", 0);
		mMax = _attrs.getAttributeIntValue(androidns, "max", 5);
	}

	@Override
	protected View onCreateDialogView() {

		View v = LayoutInflater.from(getContext()).inflate(
				R.layout.sl_slider_pref, null);

		mValueText = (TextView) v.findViewById(R.id.sl_slider_pref_txt_value);

		mSeekBar = (SeekBar) v.findViewById(R.id.sl_slider_pref_seekbar);
		mSeekBar.setOnSeekBarChangeListener(this);

		if (shouldPersist()) {
			mValue = getPersistedInt(mDefault);
		}

		mSeekBar.setMax(mMax);
		mSeekBar.setProgress(mValue);

		return v;
	}

	@Override
	protected void onBindDialogView(View view) {
		super.onBindDialogView(view);

		mSeekBar.setMax(mMax);
		mSeekBar.setProgress(mValue);
	}

	@Override
	protected void onSetInitialValue(boolean restorePersistedValue,
			Object defaultValue) {
		super.onSetInitialValue(restorePersistedValue, defaultValue);

		if (restorePersistedValue) {
			mValue = (shouldPersist()) ? getPersistedInt(mDefault) : 0;
		} else {
			mValue = (int) defaultValue;
		}
	}

	@Override
	protected void showDialog(Bundle state) {
		super.showDialog(state);

		Button positiveButton = ((AlertDialog) getDialog())
				.getButton(AlertDialog.BUTTON_POSITIVE);
		positiveButton.setOnClickListener(this);
	}

	@Override
	public void onClick(View v) {

		if (shouldPersist()) {

			mValue = mSeekBar.getProgress();
			persistInt(mValue);
			callChangeListener(mValue);
			getDialog().dismiss();
		}

	}

	//--------------------------- OnSeekBarChangedListener ---------------------------
	@Override
	public void onProgressChanged(SeekBar seekBar, int progress,
			boolean fromUser) {

		String t = String.valueOf(progress);
		mValueText.setText((mUnit == null) ? t : t.concat(" " + mUnit));
	}

	@Override
	public void onStartTrackingTouch(SeekBar seekBar) {
	}

	@Override
	public void onStopTrackingTouch(SeekBar seekBar) {
	}
	//--------------------------------------------------------------------------------

}

File: core/src/main/java/at/shockbytes/corey/common/core/util/view/SliderPreference.kt
package at.shockbytes.corey.common.core.util.view

import android.app.AlertDialog
import android.content.Context
import android.os.Bundle
import android.preference.DialogPreference
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnClickListener
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.TextView
import at.shockbytes.corey.common.core.R

/**
 * This is an own implementation of the SliderPreference, which allows the user
 * to choose between an interval from 0 - x seconds. This interval is the
 * location update interval in the TrackingService class.
 *
 * Author:  Martin Macheiner
 */
class SliderPreference : DialogPreference, OnSeekBarChangeListener, OnClickListener {

    private var seekbar: SeekBar? = null
    private var valueText: TextView? = null

    private var unit: String? = null

    private var defaultVal: Int = 0
    private var maxVal: Int = 0
    private var value: Int = 0

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        initialize(attrs)
    }

    constructor(
        context: Context,
        attrs: AttributeSet,
        defStyleAttr: Int
    ) : super(context, attrs, defStyleAttr) {
        initialize(attrs)
    }

    private fun initialize(attrs: AttributeSet) {

        // Get string value for suffix
        val unitId = attrs.getAttributeResourceValue(androidns, "text", 0)
        unit = if (unitId == 0)
            attrs.getAttributeValue(androidns, "text")
        else
            context.getString(unitId)

        defaultVal = attrs.getAttributeIntValue(androidns, "defaultValue", 5)
        maxVal = attrs.getAttributeIntValue(androidns, "max", 5)
    }

    override fun onCreateDialogView(): View {
        val v = LayoutInflater.from(context).inflate(R.layout.sl_slider_pref, null, false)

        valueText = v.findViewById(R.id.sl_slider_pref_txt_value) as TextView
        seekbar = v.findViewById(R.id.sl_slider_pref_seekbar) as SeekBar

        if (shouldPersist()) {
            value = getPersistedInt(defaultVal)
        }

        seekbar?.max = maxVal
        seekbar?.progress = value
        seekbar?.setOnSeekBarChangeListener(this)
        return v
    }

    override fun onBindDialogView(view: View) {
        super.onBindDialogView(view)

        seekbar?.max = maxVal
        seekbar?.progress = value
    }

    override fun onSetInitialValue(
        restorePersistedValue: Boolean,
        defaultValue: Any?
    ) {
        super.onSetInitialValue(restorePersistedValue, defaultValue)

        value = if (restorePersistedValue) {
            if (shouldPersist()) getPersistedInt(defaultVal) else 0
        } else {
            defaultValue as? Int ?: defaultVal
        }
    }

    override fun showDialog(state: Bundle?) {
        super.showDialog(state)

        val positiveButton = (dialog as AlertDialog)
                .getButton(AlertDialog.BUTTON_POSITIVE)
        positiveButton.setOnClickListener(this)
    }

    override fun onClick(v: View) {

        if (shouldPersist()) {
            value = seekbar?.progress ?: 0
            persistInt(value)
            callChangeListener(value)
            dialog.dismiss()
        }
    }

    // --------------------------- OnSeekBarChangedListener ---------------------------
    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
        val t = progress.toString()
        valueText?.text = t.plus(" ").plus(unit)
    }

    override fun onStartTrackingTouch(seekBar: SeekBar) {}

    override fun onStopTrackingTouch(seekBar: SeekBar) {}

    companion object {

        private val androidns = "http://schemas.android.com/apk/res/android"
    }
    // --------------------------------------------------------------------------------
}
